File: sdk/typescript/src/agent.ts
/**
 * runStrategy — minimal polling loop for an AITRAD agent.
 *
 * Polls the signal feed every `intervalMs`, deduplicates by signal id,
 * and fires `handler` for each new signal. Built for clarity not
 * throughput; when push-based market data ships (Phase 4.5) this will
 * be replaced with a WebSocket subscriber.
 */
import type { AITRADClient } from "./client.js";

export interface RunStrategyOptions {
  client: AITRADClient;
  /** ms between polls. Default 5000. */
  intervalMs?: number;
  /** Filter — "operation" (default), "strategy", "discussion", or null. */
  messageType?: string | null;
  /** Optional market filter — "us-stock", "crypto", etc. */
  market?: string;
  /** Size of the first fetch (to seed the seen-set). Default 50. */
  initialLimit?: number;
  /** Stop after N polls. Default Infinity. Mostly for tests. */
  maxIterations?: number;
  /** Hook called when the poll itself fails (network etc). No-op by default. */
  onPollError?: (err: unknown) => void;
  /** Hook called when handler throws. No-op by default. */
  onHandlerError?: (err: unknown, signal: Record<string, unknown>) => void;
}

/**
 * Poll the signal feed and dispatch new signals to `handler`.
 *
 *   await runStrategy(sig => console.log(sig), {
 *     client, intervalMs: 5000,
 *   });
 */
export async function runStrategy(
  handler: (signal: Record<string, unknown>) => void | Promise<void>,
  opts: RunStrategyOptions,
): Promise<void> {
  const {
    client,
    intervalMs = 5_000,
    messageType = "operation",
    market,
    initialLimit = 50,
    maxIterations = Number.POSITIVE_INFINITY,
    onPollError,
    onHandlerError,
  } = opts;

  const seen = new Set<number | string>();

  const listArgs: { limit: number; messageType?: string; market?: string } = {
    limit: initialLimit,
  };
  if (messageType) listArgs.messageType = messageType;
  if (market) listArgs.market = market;

  // Pre-seed `seen` with the current head so the bootstrap fetch doesn't
  // fire the handler against signals that already existed.
  const bootstrap = await client.listSignals(listArgs);
  for (const sig of _signalsFrom(bootstrap)) {
    const sid = _idOf(sig);
    if (sid !== undefined) seen.add(sid);
  }

  const pollArgs = { ...listArgs, limit: 20 };

  let iters = 0;
  while (iters < maxIterations) {
    iters += 1;

    let page: unknown;
    try {
      page = await client.listSignals(pollArgs);
    } catch (err) {
      onPollError?.(err);
      await _sleep(intervalMs);
      continue;
    }

    for (const sig of _signalsFrom(page)) {
      const sid = _idOf(sig);
      if (sid === undefined || seen.has(sid)) continue;
      seen.add(sid);
      try {
        await handler(sig);
      } catch (err) {
        onHandlerError?.(err, sig);
      }
    }

    if (iters < maxIterations) {
      await _sleep(intervalMs);
    }
  }
}

// ── helpers ───────────────────────────────────────────────────────────────

function _signalsFrom(payload: unknown): Record<string, unknown>[] {
  if (Array.isArray(payload)) return payload as Record<string, unknown>[];
  if (payload && typeof payload === "object") {
    const obj = payload as Record<string, unknown>;
    for (const key of ["signals", "items", "results", "data"]) {
      const v = obj[key];
      if (Array.isArray(v)) return v as Record<string, unknown>[];
    }
  }
  return [];
}

function _idOf(sig: Record<string, unknown>): number | string | undefined {
  const id = sig.signal_id ?? sig.id;
  if (typeof id === "number" || typeof id === "string") return id;
  return undefined;
}

function _sleep(ms: number): Promise<void> {
  if (ms <= 0) return Promise.resolve();
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: sdk/python/aitrad/agent.py
"""agent.py — minimal helpers for writing an AITRAD agent loop.

run_strategy() is the simplest useful loop: poll the signal feed every
N seconds, deduplicate, and invoke a handler for each new signal. Built
for clarity not throughput — when push-based market data ships (Phase 4.5)
this will be replaced with a WebSocket subscriber.
"""
from __future__ import annotations

import time
from typing import Callable, Iterable

from aitrad.client import AITRADClient


def run_strategy(
    handler: Callable[[dict], None],
    *,
    client: AITRADClient,
    interval: float = 5.0,
    message_type: str | None = "operation",
    market: str | None = None,
    initial_limit: int = 50,
    max_iterations: int | None = None,
) -> None:
    """Poll the signal feed and dispatch new signals to ``handler``.

    Args:
        handler:        called once per new signal with the raw signal dict.
                        Exceptions inside ``handler`` are logged-and-skipped
                        so a buggy strategy doesn't kill the loop.
        client:         authenticated AITRADClient.
        interval:       seconds between polls.
        message_type:   "operation" (default), "strategy", "discussion", or None.
        market:         optional market filter (e.g. "us-stock", "crypto").
        initial_limit:  size of the first fetch (subsequent polls always 20).
        max_iterations: stop after N poll cycles (None = forever). Mostly for tests.
    """
    seen: set[int] = set()
    # Pre-seed `seen` with the current head so the first poll doesn't fire
    # the handler against everything that already existed.
    bootstrap = client.list_signals(
        limit=initial_limit,
        message_type=message_type,
        market=market,
    )
    for sig in _signals_from(bootstrap):
        sid = sig.get("signal_id") or sig.get("id")
        if sid is not None:
            seen.add(sid)

    iters = 0
    while max_iterations is None or iters < max_iterations:
        iters += 1
        try:
            page = client.list_signals(
                limit=20, message_type=message_type, market=market,
            )
        except Exception:  # noqa: BLE001 — keep loop alive across transient errors
            time.sleep(interval)
            continue

        for sig in _signals_from(page):
            sid = sig.get("signal_id") or sig.get("id")
            if sid is None or sid in seen:
                continue
            seen.add(sid)
            try:
                handler(sig)
            except Exception:  # noqa: BLE001 — buggy handler shouldn't kill the loop
                pass

        time.sleep(interval)


def _signals_from(payload: dict | list) -> Iterable[dict]:
    """The /api/signals/feed endpoint has occasionally shipped two shapes;
    this normalizer absorbs both so the loop doesn't care."""
    if isinstance(payload, list):
        return payload
    if isinstance(payload, dict):
        for key in ("signals", "items", "results", "data"):
            v = payload.get(key)
            if isinstance(v, list):
                return v
    return []
